_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,449 @@
+	let i = 0;
+	const entries = stringToFloat32(string, i);
+  i += 2;
+	const array = [];
+	for (let arrayIndex = 0; arrayIndex < entries; arrayIndex++) {
+    const object = {};
+		for (const { name, ty } of fields) {
+			let value;
+			switch (ty) {
+				case "string":
+					const length = stringToFloat32(string, i);
+					i += 2;
+					value = String.fromCharCode(
+						...Array.from({ length }, (_, j) => string.charCodeAt(j + i)),
+					);
+					i += length;
+					break;
+				case "number":
+					value = stringToFloat32(string, i);
+					i += 2;
+					break;
+				case "boolean":
+					value = string.charCodeAt(i) === 0;
+					i++;
+					break;
+			}
+			Reflect.set(object, name, value);
+		}
+    array.push(object);
+	}
+	return array;
+}
+```
+
+Here we're using another part of reflection `Reflect.set` which allows us to set a property of an existing object based on a string key (`name`).
+
+### Problems with reflection in JavaScript
+The `arrayFromString` function assumes that the caller has passed a standard array where every object has the same type. `Reflect.get` will fail at runtime as if the property doesn't exist. The downside of this is that both the dynamic property lookup and null check is expensive at runtime.
+
+One way to catch property errors ahead of time is to us a type system on top of JavaScript like TypeScript.
+
+## Writing a Rust procedural macro
+
+To get started, if you are not already in a cargo project you can create one with `cargo new <name>` command. Before we start generating code we should declare a trait as a target for our macros output.
+
+### The BinarySerializable trait
+First, we need to create a `trait` which describes the requirements for serializing and deserializing objects - let's call this the `BinarySerializable` trait. Serialization will require adding data to a buffer. Deserialization will require pulling from an iterator (which iterates over bytes of a serialized buffer) and producing a term of `Self`. *We will assume the format is well-formed and panic at runtime rather than proper handling with `Result` when deserializing.*
+
+```rust
+pub trait BinarySerializable {
+    fn serialize(self, buf: &mut Vec<u8>);
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self;
+}
+```
+
+We can implement the `BinarySerializable` trait for the primitives that will be in our structures:
+```rust
+impl BinarySerializable for bool {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        buf.push(self as u8)
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        iter.next().unwrap() == (true as u8)
+    }
+}
+
+impl BinarySerializable for u64 {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        buf.extend_from_slice(&self.to_le_bytes());
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        let mut buf = [0; u64::BITS as usize / 8];
+        // @ben this won't work right? You're pulling a single byte? Don't you need 8 bytes?
```

Comment:
Yes it does work because the callback to [fill_with](https://doc.rust-lang.org/std/primitive.slice.html#method.fill_with) is called for each slot in the `buf` array. In this case it is called 8 times (64/8), each time mutating and advancing the `iter` iterator. See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3fc9d774744d7c1040821231f4c22957

Haven't used the function before and maybe it is little unuitive and confusing it you miss the characters `|| ` to show the argument is actually a function. IMO the function should be named `fill_using` 😄